lib/src/Pos/Diffusion/Full.hs

```diff
@@ -97,6 +97,10 @@ diffusionLayerFull networkConfig lastKnownBlockVersion transport mEkgNodeMetrics
 
         -- Timer is in microseconds.
         keepaliveTimer :: Timer <- newTimer 20000000
+        -- Retry counter.  It is reset when a new block header arrives in
+        -- @`handleBlockHeaders`@ and it is incremented when
+        -- @`dnsSubscriptionWorker`@ fails to subscribe to any node.
+        retryCounter <- newMVar 0
```

Comment:
Is it right to reset it when a block comes in? Unlike the keepalive timer, I think this can be private to the subscription worker: reset it when a subscription is established.

Reply:
There's no confirmation message back to the subscriber from `subscriptionListener`. Is there some other way I detect that a subscription was established?

Reply:
Aha, that's a shame... there is a confirmation at a lower level: a reverse-direction network-transport connection is established via a little handshake. But the conversation action is run right away rather than waiting for that to happen.

Actually, there is this static peer data convention: a conversation doesn't run until the remote node's "peer data" (in this case, `VerInfo`) has been delivered. The idea is that you can choose what to do based on the version the peer reports. So in fact, the subscription conversation handler won't run until a connection has been established!

Reply:
Sorry to appear to contradict myself, I was just thinking through it as I wrote. Resetting the counter as the first action in the subscription conversation would work. If it times out / fails for some other reason, the action won't run.

Reply:
But this neglects the case of connecting to a relay which happens to be at capacity. There, the connection will appear to go through, and then close up immediately, so the retry counter would be set back to 0. That could be a problem.

Reply:
ok, I will check how this works tomorrow morning.  I've seen the handshake, I think it's in `nodeDispatcher`. Maybe there is a way to differentiate the two cases.

Reply:
Just throwing an idea around: we could also do it based on time connected. At the start of the conversation, record the current time. When the connection is lost, grab the time and take the difference with the connection start time. The bigger this number, the shorter the delay before reconnecting, probably subject to an arbitrary threshold... if it lasted more than `n` seconds, try again right away. Otherwise, exponential backoff until at least one of them lasts more than `n` seconds.

Reply:
This handles the case in which the relay boots the client because it's at capacity. That conversation will last only a few seconds at most.

Reply:
Yes, this will prevent the case Neil had in mind: spinning through the nodes and failing shortly.